Tools/flatpak/gather-r-package-info.R

```diff
@@ -20,63 +36,187 @@ giveOrderedDependencies <- function()
-		dplyr			= { deps <- append(deps, c("plogr") ) },
-		mice			= { deps <- append(deps, c("mitml") ) }
-      )
-
-      pkgDeps[[curPkg]] <- deps
-      pkgs              <- append(pkgs, deps, i)
+      download.file(url=tarballUrl, destfile=destFile);
+      succes <- TRUE;
+    })
+
+    if(!succes)
+      stop("Failed to get pkg sources for '", pkgName, "' from github for repo: '", repo, "' and commit '", commit, "'")
+  }
+
+  return(destFile)
+}
+
+downloadPkgCran <- function(curPkg)
+{
+  version <- getPkgVersion(curPkg)
+  filePkg  <- paste0(curPkg, "_", expEnv[[curPkg]],".tar.gz")
+  destFile <- paste0(destDir, filePkg)
+
+  downloaded[[curPkg]] <- list(version=version, fileName=filePkg, destFile=destFile, downloadUrl=paste0("http://static.jasp-stats.org/RPkgs/", filePkg), sha256='')
+  
+  if(!file.exists(destFile))
+  {  
+    cat(paste0("File ",destFile, " does not exist yet, try to get it from CRAN.\n"))
+    curVer   <- paste0(CRAN, 'src/contrib/', filePkg)
+    oldVer   <- paste0(CRAN, 'src/contrib/Archive/', curPkg, '/', filePkg)
+    succes   <- FALSE
+
+    cat(paste0('For pkg ', curPkg, ' trying download from "',curVer,'"\n'))
+
+    tryCatch(error=function(e) e, exp=
+    {
+      download.file(url=curVer, destfile=destFile);
+      succes <- TRUE;
+    })
+    
+    if(!succes)
+    {
+      cat(paste0('Download failed, now trying: "',oldVer,'"\n'))  
+      tryCatch(error=function(e) e, exp={
+        download.file(url=oldVer, destfile=destFile);
+        succes <- TRUE;
+      })
     }
+
+    if(!succes)
+      stop("Both downloads failed...")
+  }
+ 
+  sha256sumOutput                  <- system2('sha256sum', args=destFile, stdout=TRUE)
+  downloaded[[curPkg]]$sha256      <- strsplit(sha256sumOutput, ' ')[[1]][[1]]
+
+  return(destFile)
+
+}
+
+
+downloadPkg <- function(pkgName)
+{
+  if(is.null(specials)) defineSpecials()
+
+  filename <- "???"
+
+  #check if special case then go to github
+  filename <- ifelse(
+    exists(pkgName, where=specials, inherits=FALSE), 
+    yes = downloadPkgGithub(pkgName), 
+    no  = downloadPkgCran(pkgName)
+  )
+
+  #github doesnt need to check if file downloaded because it is just a few
+  #otherwise try to get it from Cran if it isnt in pkg-sources yet
+
+  return(filename)
+}
+
+getDeps <- function(pkgName)
+{
+  if(pkgName %in% base_pkgs)
+    return(c())
+
+  if(exists(pkgName, where=pkgDeps, inherits=FALSE)) #Avoid double work because time is precious
+  {
+    cat(paste0('Dependencies for pkg "', pkgName, '" requested more than once...\n'))
+    return(pkgDeps[[pkgName]]) #in case someone wants em?
+  }
+
+  #make sure the source pkg is in the pkg-source folder:
+  pkgFile <- downloadPkg(pkgName)
+
+  deps    <- remotes::dev_package_deps(pkgFile, repos = NULL, type = "source")$package
+  deps    <- deps[!(deps %in% base_pkgs)] #remove base pkgs
+
+  pkgDeps[[pkgName]] <- deps
+
+  padding <- paste(rep("", 30 - stringr::str_length(pkgName)), sep=' ', collapse=' ')
```

Comment:
```suggestion
  padding <- strrep(" ", 29 - nchar(pkgName))
```
- `strrep` rather than `paste(rep("", ...))`. It does need a non-empty string to rep, so 30 becomes 29.
- using `base::nchar` rather than `stringr::str_length` means one dependency less.